File: src/offer27.MirrorOfBinaryTree/mirror_of_binary_tree.cpp
/*
struct TreeNode {
	int val;
	struct TreeNode *left;
	struct TreeNode *right;
	TreeNode(int x) :
			val(x), left(NULL), right(NULL) {
	}
};*/
class Solution1 {
public:
    void Mirror(TreeNode *pRoot) {
        if (pRoot == nullptr)
            return ;
        if (pRoot->left == nullptr && pRoot->right == nullptr)
            return;
        TreeNode* tmp = pRoot->left;
        pRoot->left = pRoot->right;
        pRoot->right = tmp;
        if (pRoot->left)
            Mirror(pRoot->left);
        if (pRoot->right)
            Mirror(pRoot->right);
    }
};

/*
struct TreeNode {
	int val;
	struct TreeNode *left;
	struct TreeNode *right;
	TreeNode(int x) :
			val(x), left(NULL), right(NULL) {
	}
};*/
class Solution2 {
public:
    void Mirror(TreeNode *pRoot) {
        if (pRoot == nullptr)
            return ;
        if (pRoot->left == nullptr && pRoot->right == nullptr)
            return;
       stack<TreeNode*> stackNode;
       stackNode.push(pRoot);
        while(!stackNode.empty()){
            TreeNode* ptmp = stackNode.top();
            stackNode.pop();
            if (ptmp->left || ptmp->right){
                TreeNode* tmp = ptmp->left;
                ptmp->left = ptmp->right;
                ptmp->right = tmp;
            }
            if (ptmp->left)
                stackNode.push(ptmp->left);
            if (ptmp->right)
                stackNode.push(ptmp->right);
        }
    }
};

File: src/offer27.MirrorOfBinaryTree/mirror_of_binary_tree.java
//1
/**
public class TreeNode {
    int val = 0;
    TreeNode left = null;
    TreeNode right = null;

    public TreeNode(int val) {
        this.val = val;

    }

}
*/
public class Solution {
    public void Mirror(TreeNode root) {
        if(root == null){
            return;
        }
        if(root.left == null && root.right == null){
            return;
        }
        TreeNode tmp = root.left;
        root.left = root.right;
        root.right = tmp;
        Mirror(root.left);
        Mirror(root.right);
    }
}
//2
/**
public class TreeNode {
    int val = 0;
    TreeNode left = null;
    TreeNode right = null;

    public TreeNode(int val) {
        this.val = val;

    }

}
*/
import java.util.*;
public class Solution {
    public void Mirror(TreeNode root) {
        if(root == null){
            return;
        }
        if(root.left == null && root.right == null){
            return ;
        }
        Deque<TreeNode> stack = new LinkedList<>();
        stack.push(root);
        while(!stack.isEmpty()){
            TreeNode cur = stack.pop();
            if(cur.left != null || cur.right != null){
                TreeNode tmp = cur.left;
                cur.left = cur.right;
                cur.right = tmp;
            }
            if(cur.left != null){
                stack.push(cur.left);
            }
            if(cur.right != null){
                stack.push(cur.right);
            }
        }
    }
}